mitiq/collector.py

```diff
@@ -0,0 +1,263 @@
+# Copyright (C) 2020 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Defines utilities for efficiently running collections of circuits generated
+by error mitigation techniques to compute expectation values."""
+
+from copy import deepcopy
+import inspect
+from typing import Any, Callable, Dict, Iterable, List, Sequence, Tuple, Union
+
+import numpy as np
+
+import cirq
+from mitiq import QPROGRAM
+from mitiq.conversions import (
+    convert_to_mitiq,
+    UnsupportedCircuitError,
+    CircuitConversionError,
+)
+from mitiq.utils import _equal
+
+
+class Collector:
```

Comment:
Would propose that this get broken up into a series of functions. `is_batched_executor` can just stand alone as a function.

Collector can then become a function that has type 
```
def collected(executor: Callable, max_batch_size: int = 100) -> Callable[[Sequence[QProgram], List[float]]
```
The usage would then be:
```
# case where original executor is batched, i.e. takes a list of QPrograms
slow_result = executor([circs])
faster_executor = collected(executor)
fast_result = faster_executor([circs])


# case where original executor is not batched, i.e. takes a list of QPrograms
slow_result = [executor(circ) for circ in circs]

faster_executor = collected(executor)
fast_result = faster_executor([circs])
```
Thus `collected` have the same function signature as batched executors. A decorator version of `collected` could also be defined for ease of use.

The state of the `CircuitCollection` could be stored on the appropriate Factory by calling it separately.

What's nice about this approach is that it is clear that the collection is an optimization that wraps the executor. Its a neat trick that can the be applied compositionally to any executor that a user defines.